### PR TITLE
Implement global roles

### PR DESCRIPTION
**Pull request by stwalkerster**
Roles can now be either defined as a "local" role (like tool admin or checkuser), or a "global" role (like tool root or steward).

Global roles automatically apply across all domains. Local roles only apply in the domain in which they are granted.

There is no practical difference between how a global and a local role are treated - the only difference occurs when the role is granted. If it's defined as global in RoleConfiguration, the application will ensure it is saved as a global role instead of a local role. From a data storage point of view, a global role grant simply has the `userrole.domain` column set to null.

This means it is possible to live-hack local roles to apply as global roles or vice-versa by directly manipulating the database and it'd behave as you'd otherwise expect. Please don't actually do that though.

This change also introduces a "steward" global role. Stewards have the ability to grant +steward and +checkuser to others, and have the ability to set large IP range bans and global bans. Otherwise, their privset is the same as someone with the checkuser role.

### CODE REVIEW

**Comment by stwalkerster** [comment]
Post-review merging blocked by the Global bans PR.

File: includes/DataObjects/UserRole.php [hunk 1]
```diff
@@ -20,7 +20,7 @@ class UserRole extends DataObject
     /** @var string */
     private $role;
     /** @var int */
```
Comment (FastLizard4) [inline includes/DataObjects/UserRole.php:22]: Hmm, can this be removed now that there's a proper type annotation?

File: includes/Helpers/LogHelper.php [hunk 2]
```diff
@@ -91,49 +85,42 @@ public static function getRequestLogsWithComments($requestId, PdoDatabase $db, S
...
-     *
-     * @param Log $entry
-     *
-     * @return string
-     */
-    public static function getLogDescription(Log $entry)
+    public static function getLogDescription(Log $entry): string
     {
-        $text = "Deferred to ";
+        $text = 'Deferred to ';
```
Comment (FastLizard4) [inline includes/Helpers/LogHelper.php:90]: I see that you're doing this wholesale standardization of double quotes to single quotes, but there's a few files in this PR where that doesn't appear to have been done, such as `Logger.php`. Is there a reason for this? I wonder if it would be better to split off the quotes normalization into a separate PR for ease of review/consistency?

File: sql/patches/patch49-globalroles.sql [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        SIGNAL SQLSTATE '45000' SET message_text = @message_text;
+    END IF;
+
+    -- -------------------------------------------------------------------------
+    -- Developers - put your upgrade statements here!
+    -- -------------------------------------------------------------------------
+
+    ALTER TABLE userrole
+        MODIFY COLUMN domain int(10) unsigned NULL,
+        ADD COLUMN `global` tinyint(3) unsigned GENERATED ALWAYS AS (CASE WHEN domain IS NULL THEN  1 END) VIRTUAL COMMENT 'Is this globally-defined?',
```
Comment (FastLizard4) [inline sql/patches/patch49-globalroles.sql:39]: Purely out of curiosity, why length 3?

**Review by FastLizard4 — COMMENTED** [review]
Code LGTM, just some questions before approving

File: includes/DataObjects/UserRole.php (hunk 1 above)
Comment (stwalkerster) [reply includes/DataObjects/UserRole.php]: Yes, done

File: includes/Helpers/LogHelper.php (hunk 2 above)
Comment (stwalkerster) [reply includes/Helpers/LogHelper.php]: I've removed the quote changes from the PR for now

File: sql/patches/patch49-globalroles.sql (hunk 3 above)
Comment (stwalkerster) [reply sql/patches/patch49-globalroles.sql:39]: I think this is a codegen artefact.

For MySQL/MariaDB, this really doesn't matter for numeric types as they're all the same size under the hood. The number is just used to determine the column width when displayed on the CLI.

It could be a tinyint(1) as we're only ever going to have a single digit (a single bit really) in this. The contents of the column aren't even stored to disk.

**Review by FastLizard4 — DISMISSED** [review]

**Review by FastLizard4 — APPROVED** [review]
